Float histograms: implement methods for Add/Sub operations using Kahan summation

crush-on-anechka opened the pull request:
This PR addresses issue #14105 and PR #14325 maintained by @beorn7. While the issue has an assignee, I decided to contribute out of personal interest as there hasn’t been recent activity.

**Overview:**
New methods, kahanAdd and kahanSub, to the FloatHistogram replicate the logic of the existing Add and Sub methods but incorporate the Kahan summation algorithm. The methods require an initialized compensation histogram as input and return   both the modified receiving histogram and the updated compensation histogram. The resulting histograms must then be combined using the original Add/Sub methods. Usage examples can be seen in the modified TestFloatHistogramAdd and TestFloatHistogramSub test cases.
 
**Details:**
- I added two helper functions kahanSumDec and kahanSumInc (copied from promql/engine.go to avoid circular imports which handle additions and subtraction operations.
- I added the NewCompensationHistogram method. It initializes a compensation histogram which matches the schema and custom bucket layout of the receiving histogram, however all positive and negative bucket values are initialized to zero. The spans of the compensation histogram are shared with the receiving histogram. This ensures both histograms maintain identical span layouts. Sharing span slices avoids unnecessary operations, as the receiving histogram's spans are modified during add/sub operations.
- I created the kahanAddBuckets function as an enhanced version of AddBuckets. This function processes both the receiving histogram's buckets and the corresponding compensation histogram buckets. Similarly, I modified reduceResolution into kahanReduceResolution.
 
**Food for thought:**
- The h.reconcileZeroBuckets method, invoked in Add/Sub and kahanAdd/kahanSub, modifies the spans of the receiving histogram (it happens in Compact method   to be precise) in some cases. To keep the compensation histogram in sync, its span slices are explicitly resized accordingly. However, if the receiving histogram's spans are reallocated, I reassign the compensation histogram's spans to ensure they point to the correct underlying array. I observed that existing test cases show spans only decrease in length, which avoids reallocation in most cases. If span lengths can not increase, a reallocation check can be omitted.
- The Add method contains logic for processing the CounterResetHint, but the Sub method does not. Is this intentional, or could it be an oversight?
- The reduceResolution function currently performs floating-point additions when merging buckets. Incorporating the Kahan summation algorithm here would add complexity but Is it worth it?

beorn7 commented:
@KofClubs is also working on this (or maybe the work has stalled?). Just FYI.

beorn7 commented:
> The Add method contains logic for processing the CounterResetHint, but the Sub method does not. Is this intentional, or could it be an oversight?

Yes, I know. (See the TODO [here](https://prometheus.io/docs/specs/native_histograms/#binary-operators)). I'm not sure 
yet how to deal with negative counters in general (which wreaks havoc with the whole "only goes down if there is a counter reset" heuristics).

crush-on-anechka commented:
> @KofClubs is also working on this (or maybe the work has stalled?). Just FYI.

I know, I guess the work has stalled as there hasn’t been activity for a long time. If that's not the case, then I'm ok to step aside and not interfere. Otherwise I'm ready to keep on

beorn7 commented:
Thank you very much, I'll have a detailed look ASAP (which is probably not very soon, as I'm about to disappear into my holiday break). If anyone wants to review this in the meantime, be my guest.

I haven't heard anything from @KofClubs , so let's hope we don't have redundant work happening here.

beorn7 commented:
Back from my holiday break, but swamped by the backlog. I hope I'll get to this next week.

KofClubs commented:
@crush-on-anechka @beorn7 

I just received an email about @crush-on-anechka 's good work. I'm sorry that I put this work on hold due to the long office hours every day:(

I shall unassign the corresponding issue.

beorn7 commented:
Sorry, I caught a cold and got less done than planned. I need to ask for more patience. (And of course, if somebody else beats me to the review, be my guest.)

beorn7 commented:
> The Add method contains logic for processing the CounterResetHint, but the Sub method does not. Is this intentional, or could it be an oversight?

That's because I still haven't made up my mind how to treat counter resets for subtraction. I'll add that once I have an idea.

> The reduceResolution function currently performs floating-point additions when merging buckets. Incorporating the Kahan summation algorithm here would add complexity but Is it worth it?

Let's not do it for now. I don't expect the counts in the merged buckets to be hugely different, and usually we'll only add a few buckets anyway.

beorn7 reviewed (changes requested):
Thank you very much. Generally looks good.

Special thanks for all the careful thought you put into this.

It's a bit sad that we need to duplicate so much logic, but I guess it cannot be prevented most of the time. (For the few occasions where we can, I have left a comment.)

It would be good to already wire the new methods into PromQL (for addition, subtraction, and the sum aggregation). This will give us a lot more test coverage (via the PromQL tests), and we could even easily try out a few things in the PromQL tests to prove that the Kahan summation actually helps.

beorn7 commented:
A few more thoughts:

We currently don't do Kahan summation for the float binary `+` or `-` operator. That would require more plumbing in the PromQL engine, and I would also doubt it's terribly useful because it's unlikely to chain a lot of those operations in practical use.

Kahan summation is currently only used in certain aggregations, the corresponding "over time" functions and some specific functions. Many of those only act on float samples. The only aggregations and functions we will need Kahan summation for histograms for will be:
- `sum` aggregation
- `avg` aggregation
- `sum_over_time` function
- `avg_over_time` function

As far as I can see, we do not need `KahanSub` for any of those. So we actually do not need those parts.

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,6 +410,139 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
 }
 
+// kahanSumInc performs addition of two floating-point numbers using the Kahan summation algorithm.
+func kahanSumInc(inc, sum, c float64) (newSum, newC float64) {
```
done! thanks for hinting correct place for this function

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,6 +410,139 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+// KahanAdd works like Add but using the Kahan summation algorithm to minimize numerical errors.
+// It returns pointers to the updated receiving histogram
+// and a separate histogram that holds the Kahan compensation term.
+func (h *FloatHistogram) KahanAdd(other, c *FloatHistogram) (*FloatHistogram, *FloatHistogram, error) {
```
named return - done, a small fix, hope that resolves the issue.
generate the compensation histogram within KahanAdd - done. newCompensationHistogram is now unexported.

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,6 +410,139 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+		// Conservatively set the CounterResetHint to "unknown" and issue a warning.
+		h.CounterResetHint = UnknownCounterReset
+		// TODO(trevorwhitney): Actually issue the warning as soon as the plumbing for it is in place
+	}
```
moved this to adjustCounterResetForAddition. I found this name more self-explanatory

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,6 +410,139 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+		h.ZeroCount, c.ZeroCount = kahanSumInc(otherZeroCount, h.ZeroCount, c.ZeroCount)
+
+		// Ensure c.PositiveSpans and c.NegativeSpans match h.PositiveSpans and h.NegativeSpans.
+		// Reassign if the underlying arrays have been reallocated; otherwise, resize to match lengths.
```
thanks for hinting that this is cheaper than the comparisons, in this case it's more neat and safe indeed

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -461,6 +594,83 @@ func (h *FloatHistogram) Sub(other *FloatHistogram) (*FloatHistogram, error) {
+		h.ZeroCount, c.ZeroCount = kahanSumDec(otherZeroCount, h.ZeroCount, c.ZeroCount)
+
+		// Ensure c.PositiveSpans and c.NegativeSpans match h.PositiveSpans and h.NegativeSpans.
+		// Reassign if the underlying arrays have been reallocated; otherwise, resize to match lengths.
```
this one is outdated now, I guess it was related to KahanSub which we omitted

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -1325,6 +1535,138 @@ func addBuckets(
+	spansA []Span, bucketsA []float64,
+	spansB []Span, bucketsB []float64,
+	bucketsC []float64,
+) ([]Span, []float64, []float64) {
```
done

crush-on-anechka commented on model/histogram/generic.go:
```diff
@@ -778,6 +778,130 @@ func reduceResolution[IBC InternalBucketCount](
+	targetSchema int32,
+	deltaBuckets bool,
+	inplace bool,
+) ([]Span, []float64, []float64) {
```
done

crush-on-anechka commented:
@beorn7 I fixed the issues. I also replaced Add with KahanAdd in engine.go/aggregation and corresponding functions as you mentioned, but I'm a little bit confused with how to handle errors in adding up the accumulated receiving histogram and compensation histograms (lines 3129 and 3161) since handleAggregationError demands metricName. The solution could be to store metric name in groupedAggregation so we would have access to them, but I'm not sure.

beorn7 commented:
Thanks for all the updates. I'll have a look ASAP.

beorn7 commented:
Still being caught is other things. Hopefully I'll get to this before this weeks ends.

beorn7 reviewed (changes requested):
Thank you very much for your patience and all the epic work.

This looks very good in general. Among all the comments, there are two substantial topics:
1. Handling of the zero bucket (I assume that's not yet correct, see comment for details).
2. Like for floats, we should first try to do simple average calculation and only revert to incremental average calculation if we run into overflow problems.

(About the 2nd point: Without Kahan summation, the incremental average calculation also helps to avoid the problem of adding a huge number (the running sum) to a small number (a new sample). However, Kahan summation solves that in a different way, so now the incremental calculation is only needed if the running sum would overflow float64. If that's not the case, it's just additional work _and_ additional inaccuracies. See  #14413 for some historical context.)

We need to address (1) in this PR, but you could leave TODOs about (2) if you feel it's too much for this PR.

Thanks to wiring the Kahan summation into PromQL already, we get some test coverage for free. However, let's add some more tests for histograms to make sure everything works as expected. All of these tests should be done via the PromQL testing framework, so they are easy to write (just using PromQL, no Go code required). You could take inspiration from float tests, e.g. https://github.com/prometheus/prometheus/blob/61aa82865d9c8474393bbbdcd539c58f63ba514f/promql/promqltest/testdata/aggregators.test#L612C1-L617C1 would not work without Kahan summation, https://github.com/prometheus/prometheus/blob/61aa82865d9c8474393bbbdcd539c58f63ba514f/promql/promqltest/testdata/aggregators.test#L576C1-L585C1 exercises the encremental avg calculation. 

It would also be good to construct a test that exposes the (suspected) bug with the zero bucket, see point (1) above.

crush-on-anechka commented:
@beorn7 Thanks for the feedback! I'll analyze the information and update the PR asap

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+
+	if h.UsesCustomBuckets() {
+		h.PositiveSpans, h.PositiveBuckets, c.PositiveBuckets = kahanAddBuckets(
+			h.Schema, h.ZeroThreshold, false, hPositiveSpans, hPositiveBuckets, otherPositiveSpans, otherPositiveBuckets, cPositiveBuckets)
```
I second that! Is there a convention on max line length? BTW I could also refactor Add and Sub (at least those two, as they share similar blocks with KahanAdd) to match the coding style if you'd like.

beorn7 replied:
No strict max line length. Just use your best judgement.

(Note that the style quirk is not about the line length, but about putting the closing parenthesis on the next line with multi line function calls.)

crush-on-anechka replied:
Thanks for clarifying, I got this! You suggest to split a long line in your later comment, that's why I asked as it's also related to coding style. I’ll go with the flow then.

crush-on-anechka replied:
fixed

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+	h.PositiveSpans, h.PositiveBuckets, c.PositiveBuckets = kahanAddBuckets(
+		h.Schema, h.ZeroThreshold, false, hPositiveSpans, hPositiveBuckets, otherPositiveSpans, otherPositiveBuckets, cPositiveBuckets)
+	h.NegativeSpans, h.NegativeBuckets, c.NegativeBuckets = kahanAddBuckets(
+		h.Schema, h.ZeroThreshold, false, hNegativeSpans, hNegativeBuckets, otherNegativeSpans, otherNegativeBuckets, cNegativeBuckets)
```
looks good, fixed

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+	}
+
+	h.PositiveSpans, h.PositiveBuckets, c.PositiveBuckets = kahanAddBuckets(
+		h.Schema, h.ZeroThreshold, false, hPositiveSpans, hPositiveBuckets, otherPositiveSpans, otherPositiveBuckets, cPositiveBuckets)
```
done

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+		otherPositiveSpans, otherPositiveBuckets = reduceResolution(
+			otherPositiveSpans, otherPositiveBuckets, other.Schema, h.Schema, false, false)
+		otherNegativeSpans, otherNegativeBuckets = reduceResolution(
+			otherNegativeSpans, otherNegativeBuckets, other.Schema, h.Schema, false, false)
```
fixed

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+
+	case other.Schema > h.Schema:
+		otherPositiveSpans, otherPositiveBuckets = reduceResolution(
+			otherPositiveSpans, otherPositiveBuckets, other.Schema, h.Schema, false, false)
```
done

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+	switch {
+	case other.Schema < h.Schema:
+		hPositiveSpans, hPositiveBuckets, cPositiveBuckets = kahanReduceResolution(
+			hPositiveSpans, hPositiveBuckets, cPositiveBuckets, h.Schema, other.Schema, false, true)
```
done

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+		hPositiveSpans, hPositiveBuckets, cPositiveBuckets = kahanReduceResolution(
+			hPositiveSpans, hPositiveBuckets, cPositiveBuckets, h.Schema, other.Schema, false, true)
+		hNegativeSpans, hNegativeBuckets, cNegativeBuckets = kahanReduceResolution(
+			hNegativeSpans, hNegativeBuckets, cNegativeBuckets, h.Schema, other.Schema, false, true)
```
done

crush-on-anechka commented on promql/functions_internal_test.go:
```diff
@@ -18,6 +18,7 @@ import (
 	"testing"
 
+	"github.com/prometheus/prometheus/util/kahansum"
 	"github.com/stretchr/testify/require"
```
done

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+// KahanAdd works like Add but using the Kahan summation algorithm to minimize numerical errors.
+// It returns pointers to the updated receiving histogram
+// and a separate histogram that holds the Kahan compensation term.
+func (h *FloatHistogram) KahanAdd(other, c *FloatHistogram) (newH *FloatHistogram, newC *FloatHistogram, err error) {
```
done, thanks!

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -410,13 +387,79 @@ func (h *FloatHistogram) Add(other *FloatHistogram) (*FloatHistogram, error) {
+	}
+
+	if !h.UsesCustomBuckets() {
+		otherZeroCount := h.reconcileZeroBuckets(other)
```
I wanted to avoid excessive calls to reconcileZeroBuckets() since it invokes Compact(), which is known to be an expensive operation. However, I didn’t analyze it in depth indeed. This one turned out to be a bit tricky after all.

The reconcileZeroBuckets() function modifies buckets in some cases — specifically, it eliminates zero-value buckets. This disrupts the compensation histogram, which initially contains only zeroed buckets. As a result, the bucket layout of the compensation histogram no longer matches that of the receiving histogram.

Here’s my thoughts and the solution I came up with (I updated the KahanAdd method and added some test cases):

The Compact() method is called when the zero threshold changes, which happens when adding two histograms with different zero thresholds. Compact() performs two key operations:
- It removes buckets whose values are exactly zero.
- It eliminates buckets that are now absorbed by the increased zero threshold, incrementing the zero count accordingly.

Since Compact() only decreases the number of buckets (never increases them), we can determine whether it was called by comparing the lengths of the receiving histogram's buckets and the compensation histogram's buckets. If the receiving histogram’s bucket slice is shorter than the compensation histogram’s, it means that zero-value buckets were removed.

To align the compensation histogram with the modified receiving histogram, we first identify and store the indices of zero-value buckets in the receiving histogram before calling reconcileZeroBuckets(). We then remove the corresponding values at those indices from the compensation histogram’s buckets.

After this adjustment, if the receiving histogram’s bucket slice is still shorter than the compensation histogram’s, it indicates that additional buckets were absorbed by the increased zero threshold. In this case, we further trim the leftmost (closest to zero) values from the compensation histogram’s bucket slice.

Again, this is a corner case, but the solution introduces additional overhead by requiring traversal of positive and negative buckets, storing indices, and comparing bucket slice lengths. To mitigate this overhead on every iteration, I introduced a boolean variable, shouldAdjustCompensationBuckets, which is set based on other.ZeroThreshold > h.ZeroThreshold — the only condition that triggers Compact().

What do you think of this @beorn7 ?

beorn7 replied:
After some considerations, I'm not sure that your approach works. It also seems quite complicated (which makes it harder to find out if it works or not).

Here are my thoughts:

You are right that just calling `c.reconcileZeroBuckets` is a problem because the compensation histogram may have unpopulated buckets, which would change the behavior of `reconcileZeroBuckets` and which will also be compacted away.

However, I don't think that `shouldAdjustCompensationBuckets = other.ZeroThreshold > h.ZeroThreshold` is a safe assumption. If `other.ZeroThreshold` is smaller than `h.ZeroThreshold`, `other.ZeroThreshold` is increased. Ideally, increasing it to exactly `h.ZeroThreshold` works. However, if `h.ZeroThreshold` happens to be inside a populated bucket of `other`, `other.ZeroThreshold` has to be increased even more – until it has reached a value that is not inside a populated bucket in either histogram. If that happens, then also `h` needs to be adjusted. But `shouldAdjustCompensationBuckets` would still be `false`.

So the only safe assumption is `shouldAdjustCompensationBuckets = other.ZeroThreshold != h.ZeroThreshold`.

However, the compaction of `h` comes with another caveat: In principle, `h` might have empty buckets corresponding to non-empty buckets in `c`. In that case, `c` would have a bucket that isn't even represented in `h` anymore.

It appears to me that an adjustment of `c` after compaction of `h` is impossible. (Also note that sharing the span slices between `h` and `c` is actually not safe because of this possible compaction, which modifies the span slices in-place, so `c` would have non-compacted buckets but spans that do not match the buckets. This part, at least, is easily solvable: We simply have go hive the compensation histogram its own span slices (see other comment).)

One way out might be to implement a "three-way zero bucket reconciliation", maybe called `h.kahanReconcileZeroBuckets(other, c *FloatHistogram)`. It could modify `h` and `c` in parallel and ensure the shared layout (we could then keep sharing the spans). But this would require a re-implementation of the whole algorithm of reconciling zero buckets with those two histograms in parallal.

Another idea would be to drop the idea of having a shared layout between `h` and `c`. Just start with `c` as an empty histogram and let `kahanAddBuckets` add buckets to `c` as needed. This means that `kahanAddBuckets` also needs a `spansC []Span` argument. There would still be the problem of the "three way zero bucket reconciliation", but maybe this could be done by calling `reconcileZeroBuckets` in succession until all is consistent.

(Side note: I wouldn't worry too much about the performance impact of the `reconcileZeroBuckets` calls, as the case where it is actually needed is pretty rare.)

crush-on-anechka replied:
ah yes, now I see that `shouldAdjustCompensationBuckets = other.ZeroThreshold != h.ZeroThreshold` is indeed the correct assumption.

Additionally, I've realized that `Compact` does not always remove all zero-valued buckets in certain cases. Am I right? If that's true, adjusting `c` after compacting `h` may indeed be impossible. Otherwise storing the indices of zero valued buckets of `h` (before compacting) and deleting `c.buckets` at those indices might work.

Among the two ideas you suggested, to me compacting `h` and `c` in parallel looks more straightforward. This way, we only introduce `c.buckets` without messing around with multiple spans. We can still use shared spans since `c.spans` are explicitly reassigned after `reconcileZeroBuckets`, and `kahanAddBuckets` remains unchanged.

Moreover, `c.spans` and `c.schema` are reassigned in every iteration of `kahanAdd`, yet they are not actually used within it. They only need to be in sync when calling `Add` to finalize the summation loop by adding up the receiving histogram and accumulated compensation. If we move their reassignment outside `kahanAdd` and perform it once, we could reduce redundant steps. I ran benchmarks and found no meaningful performance difference so it's purely a matter of code structure. What do you think?

beorn7 reviewed (commented):
Here my thoughts. The zero bucket handling is really tough… WDYT?

beorn7 commented:
To go even further down the rabbit hole:

When reconciling zero buckets, we might be merging buckets into the zero bucket.

Even when ignoring zero buckets, we will still merge buckets if we sum up histograms of different resolution.

If we merge buckets, we add numbers. And we might need to add a very small to a very large number. So we should _also_ use Kahan summation in this case, given that we have the compensation value available anyway.

However, now we don't add one number to another one with its compensation value, we add two numbers with each having their own compensation term. Let's say bucket b1 with c1, and bucket b2 with c2. We need to check the math, but I guess it will be something like first doing `b1, c1 = kahansum.Inc(b2, b1, c1)` and then `c1 += c2`.

crush-on-anechka commented:
@beorn7  regarding the rabbit hole - I see where you're coming from, I'll take care of it

crush-on-anechka commented:
@beorn7 in three latest commits:

1. Handling zero bucket and compacting.
I refactored `reconcileZeroBuckets` and all related functions, passing the compensation histogram into it. Since `reconcileZeroBuckets`, `zeroCountForLargerThreshold`, `trimBucketsInZeroBucket` and `compactBuckets` are called infrequently and only within the same package, I reviewed all their call sites and explicitly passed `nil` for the compensation histogram where it is not needed. This avoids duplicating these functions as Kahan-specific variants. However, `Compact` is widely used across the codebase, so I introduced `kahanCompact`.

2. Merging buckets into the zero bucket (`zeroCountForLargerThreshold`) and merging during resolution reduction (`kahanReduceResolution`) are now implemented following the approach you described.

beorn7 commented:
This sounds wonderful. I hope it all works out as planned. I'll have a look ASAP (which might need another week or so because we are going into KubeCon week, where I'll be very busy). In any case, thank you once more.

crush-on-anechka commented:
Sure! I'll use this time to check on the things you mentioned earlier that I haven't gotten to yet

beorn7 commented:
The idea is to only switch to incremental calculation once we would everflow float64. An imprecise average is still better than the Inf or NaN we get when overflowing. Check how it is done for averaging the simple float samples.

crush-on-anechka commented:
After a series of experiments, I found that both incremental and non-incremental average calculation involving large-magnitude values do not behave as expected for simple floats also.

Summing numbers of drastically different magnitudes leads to precision loss. Kahan helps reduce errors, but it still relies on basic `+` and `-` operations, which themselves are subject to rounding when the values differ too much in scale.

For example we try to add `2.08169134891958e107` and `1.13729473572312e93`.
Converted to the same exponent:
`1.13729473572312e93 ≈ 0.0000000000000113729473572312e107`
Since float64 holds ~16 significant digits, the smaller number is rounded to `0.00000000000001e107`.

While reviewing the current summing algorithm, I noticed some potential issues and experimented with an updated version, nevertheless both of them don't work as expected in certain cases:

```
input: 1e+220 45 -1e+220
expected avg = 15

actual avg (incremental - current version) = 0
45 is lost - absorbed into 1e+220 and wasn't captured by the compensation term.

actual avg (incremental - updated version) = -2.0901902125471323e+203
on the final step (calculating the delta):
-1e+220 - 5e+219 = -1.4999999999999999e+220 (should be -1.5e+220)
This triggers a huge compensation term: -1.2541141275282797e+204, which corrupts the result.
```

```
input: 4.671249346736492e30 1e17 54 -1e17 -4.671249346736492e30 0
expected avg = 9

actual avg (incremental - current version) = 5.8640620148053336e+13
actual avg (incremental - updated version) = -7.237358535746467e+13
Although the values balance out by the end, a massive compensation term (13 orders of magnitude) remains and is applied to the final result, producing incorrect average.
```

```
input: 6.6666666666666666e107 1e53 45 -6.6666666666666666e107 -1e53
expected avg = 9

actual avg (incremental - current version) = -2e+52
1e53 and 45 are lost, so basically only the numbers with a magnitude around 1e107 are effectively taken into account, resulting in a final mean of 0.

actual avg (incremental - updated version) = 0
actual avg (non-incremental) = 0
45 is lost - absorbed into a compensation term ~52 orders of magnitude higher.
```

As you can see, non-incremental calculation also fails when the input values span three different orders of magnitude. If we only had values around ~100 orders of magnitude and others below ~14 orders of magnitude, the calculation would work: the main sum would handle the large numbers, while the compensation would account for the small ones. However, introducing values in the middle range (around ~50 orders of magnitude) corrupts the calculation.

TL;DR:
Are there any constraints or assumptions around the input data? Mixing values with such a wide range of magnitudes (e.g., 1e100, 1e50, and 45) is impossible to handle accurately using float64, whether in incremental or non-incremental calculation. Most likely, my test cases are much more extreme than what happens with real data. So far, I have more questions than answers — sorry about that!

beorn7 commented:
Thanks for your thoughts. 

There aren't really any assumptions about the input values. Users can do whatever they want. However, there is no expectation that we can solve all floating point precision issues. Even Kahan summation can take us only so far.

I'm not an expert in floating point arithmetics, but your line of argument that providing values from many different orders of magnitude will eventually break even Kahan summation, sounds plausible. 

I usually try to imagine Kahan summation as "getting something like float128 precision even on a float64 FPU" – at somewhat higher cost, which doesn't really make a dent in the PromQL case because fetching all those numbers from the TSDB is much more expensive that the additional summation we have to do for Kahan. But even float128 will eventually run into precision issues.

About your examples:

`1e+220 45 -1e+220` should just work fine with the current implementation. In fact, I tried it out, and it worked:

```
# Test summing and averaging extreme values.
clear

load 10s
	data{test="ten",point="a"} 1e+220
	data{test="ten",point="b"} 45
	data{test="ten",point="c"} -1e220

eval instant at 1m sum(data{test="ten"})
	{} 45

eval instant at 1m avg(data{test="ten"})
	{} 15
```

Where did you get your numbers from?

If you add other values with an order of magnitude in between, like the following, I can also see the problem:

```
load 10s
	data{test="ten",point="a"} 1e+220
	data{test="ten",point="b"} 45
	data{test="ten",point="c"} -1e220
	data{test="ten",point="d"} 1e22
	data{test="ten",point="e"} -1e22
```

But as said, I think that is expected. However, I expect the outcome to be close to zero here, not any gigantic number. Again, I don't know how you got the numbers you quoted above.

crush-on-anechka commented:
Apologies for the lack of clarity in my previous message. I’ve been experimenting with the avg_over_time function, and here’s how I set up the test:
```
clear
load 10s
  metric 1e+220 45 -1e+220

eval instant at 55s avg_over_time(metric[1m])
  {} 15
```
To force the incremental calculation path, I explicitly set `incrementalMean = true`.

When I referred to the “updated version” of the algorithm, I meant this one:
```
var delta float64
delta, kahanC = kahansum.Inc(-mean, f.F, kahanC)
delta /= count
kahanC /= count / (count - 1)
mean += delta
```
This is in contrast to the current version:
```
correctedMean := mean + kahanC
mean, kahanC = kahansum.Inc(f.F/count-correctedMean/count, mean, kahanC)
```

beorn7 commented:
(Note for the reader to not lose track: We are currently discussion the average calculation for simple floats, not for histograms. We just stumbled upon problems with floats while trying to figure out how to do histograms.)

The current non-incremental version works:

```
load 10s
  metric14 1e+220 45 -1e+220

eval instant at 55s avg_over_time(metric14[1m])
  {} 15
```

The problem is if the test is performed `at 1m` because then the first value is not included (anymore, since v3). This is actually a problem in existing tests. I've created https://github.com/prometheus/prometheus/pull/16566 to fix that, please have a look. (We already had tests further down testing the above scenario.)

Even the incremental versions should give "correct" results within floating point accuracy, of course. I'm not sure if our current "hybrid incremental/Kahav avg calculation" is ideal (or maybe even correct). I found a bit of discussion on [Stackoverflow](https://stackoverflow.com/questions/61665473/is-it-beneficial-for-precision-to-calculate-the-incremental-mean-average), but it fell short to provide the actual algorithm that combines Kahan and incremental avg calculation.

I think your "updated version" is not quite correct. I have a slightly different idea. Maybe with that idea in place, we can get rid of the whole "switching to incremental calculation if needed" thing. I'm working on a PR to present to you for further discussion.

beorn7 commented:
I have created https://github.com/prometheus/prometheus/pull/16569 for the (hopefully) correct combination of Kahan and incremental mean calulation.

crush-on-anechka commented:
#16569 looks tidy indeed! My thoughts:
- You’ve omitted the non-incremental calculation path, which actually performs better in cases where large positive and negative values should cancel each other out. Unfortunately, even with the Kahan summation improvement, the incremental approach can't handle these scenarios accurately due to float64 rounding limitations.
- I’m a bit puzzled because I haven’t found a test case where Kahan summation clearly outperforms direct summation in terms of accuracy. In fact, there is one example:
```
load 5s
  metric1 1e10 1e-6 1e-6 1e-6 1e-6 1e-6

eval instant at 55s avg_over_time(metric1[1m])
  {} 1.666666666666675e+09
```
This yields `1.6666666666666675e+09` with Kahan and `1.666666666666668e+09` with direct summation, but the difference is minimal and the test still passes as it's within the accepted tolerance.

I’ve written additional tests for the approach in #16569 and grouped them by behavior (see comments). `metric9`, `metric10`, and `metric11` fail due to the issue of large positive and negative values that should cancel out. While `metric9` and `metric10` produce results that are at least close to expected, `metric11` is significantly off.
```
clear
# All positive values with varying magnitudes
load 5s
  metric1 1e10 1e-6 1e-6 1e-6 1e-6 1e-6
  metric2 5.30921651659898 0.961118537914768 1.62091361305318 0.865089463758091 0.323055185914577 0.951811357687154
  metric3 1.78264e50 0.76342771 1.9592258 7.69805412 458.90154
  metric4 0.76342771 1.9592258 7.69805412 1.78264e50 458.90154
  metric5 1.78264E+50 0.76342771 1.9592258 2.258E+220 7.69805412 458.90154

# Contains negative values; result is dominated by a large-magnitude value
load 5s
  metric6 -1.78264E+50 0.76342771 1.9592258 2.258E+220 7.69805412 -458.90154
  metric7 -1.78264E+50 0.76342771 1.9592258 -2.258E+220 7.69805412 -458.90154
  metric8 -1.78264E+215 0.76342771 1.9592258 2.258E+220 7.69805412 -458.90154

# Contains negative values; large value dominates, but smaller values should cancel each other out
load 5s
  metric9 -1.78264E+215 0.76342771 1.9592258 2.258E+220 7.69805412 1.78264E+215 -458.90154
  metric10 -1.78264E+219 0.76342771 1.9592258 2.3757689E+217 -2.3757689E+217 2.258E+220 7.69805412 1.78264E+219 -458.90154

# High-magnitude values should cancel each other out, leaving a small result
load 5s
  metric11 -2.258E+220 -1.78264E+219 0.76342771 1.9592258 2.3757689E+217 -2.3757689E+217 2.258E+220 7.69805412 1.78264E+219 -458.90154

eval instant at 55s avg_over_time(metric1[1m])
  {} 1.6666666666666675e+09

eval instant at 55s avg_over_time(metric2[1m])
  {} 1.67186744582113
  
eval instant at 55s avg_over_time(metric3[1m])
  {} 3.56528E+49
  
eval instant at 55s avg_over_time(metric4[1m])
  {} 3.56528E+49
  
eval instant at 55s avg_over_time(metric5[1m])
  {} 3.76333333333333E+219
  
eval instant at 55s avg_over_time(metric6[1m])
  {} 3.76333333333333E+219
  
eval instant at 55s avg_over_time(metric7[1m])
  {} -3.76333333333333E+219
  
eval instant at 55s avg_over_time(metric8[1m])
  {} 3.76330362266667E+219
  
eval instant at 55s avg_over_time(metric9[1m])
  {} 3.228571428571429e+219
  
eval instant at 55s avg_over_time(metric10[1m])
  {} 2.5111e+219
  
eval instant at 55s avg_over_time(metric11[1m])
  {} -44.848083237
```

beorn7 commented:
Thank you. I still have to process your latest response with the depth it deserves. However, I'm still caught between several events at work and in my usual life, but I will hopefully get to it next week.

beorn7 commented:
Thanks for the additional test cases.

I have tried them out, and for all that fail with #16569, they also fail before #16569, just with more or less differing values. Thus, I'm still not convinced that the incremental approach with "correct" Kahan summation as in #16569 is really performing worse than direct Kahan summation with the one division at the end (what I call "simple mean calculation").

I will incorporate your test cases in #16569 and discuss them over there. (Hopefully I'll get to it tomorrow.) We can continue the detailed discussion over there.

About your other comment:

> I’m a bit puzzled because I haven’t found a test case where Kahan summation clearly outperforms direct summation in terms of accuracy.

Not quite sure what you are referring to here. We have two dimensions here:
1. Kahan summation vs. regular summation.
2. Incremental mean calculation vs. simple mean calculation.

What do you mean by "direct summation"? There are plenty of test cases already in the test data that would fail without Kahan summation.

beorn7 commented:
For `metric10`, you said the expected value is 2.5111e+219. But again, this series is completely dominated by 2.258e+220, so the correct average should just be 2.258e+220 / 9 = 2.5088888888888888e+219. Interestingly, this is what happens after #16569 (and also with the Python statistics package), but before #16569, we get something fairly wrong (3.225714285714286e+219).

crush-on-anechka commented:
You're right — I suspect my earlier results were off due to inaccuracies in the averaging method I used. Using Python's statistics is a good idea, I will stick with it going forward.
Regarding your comment on metric10: how did you get 3.225714285714286e+219? It's not arbitrary but aligns with the value from metric9 (2.258e+220 / 7). It looks like the metric10 test case may have been trimmed from 9 to 7 entries. I actually get 2.5088888888888888e+219 in my version before #16569.

crush-on-anechka commented:
Regarding my comment on "Kahan summation outperforming direct summation" — what I actually meant was "Kahan summation vs. regular summation."

Since I was working on the `funcAvgOverTime` function and we encountered the incremental vs. non-incremental issue, my tests primarily focused on that. However I also ran the same algorithm without the Kahan compensation tracking and the results turned out to be identical.

I then tried to come up with scenarios where Kahan compensation would make a noticeable difference, but I couldn't find any clear-cut cases. This was just a side experiment driven by curiosity, and if there are already tests confirming the effectiveness of the Kahan implementation, I'm happy to disregard my earlier comment.

beorn7 commented:
> -1.78264E+219 0.76342771 1.9592258 2.3757689E+217 -2.3757689E+217 2.258E+220 7.69805412 1.78264E+219 -458.90154

beorn7 commented:
> Regarding your comment on metric10: how did you get 3.225714285714286e+219? It's not arbitrary but aligns with the value from metric9 (2.258e+220 / 7). It looks like the metric10 test case may have been trimmed from 9 to 7 entries. I actually get 2.5088888888888888e+219 in my version before #16569.

You are right. I must have left something out of `metric10` when I tried it with the pre-#16569 state. The test actually works fine. I'll remove the comment.

Oven in https://github.com/prometheus/prometheus/issues/16714, we also got some more evidence that the direct mean calculation is doing better than the incremental mean calculation. My plan is now to combine what we had before #16569 with the improved incremental mean calculation.

beorn7 commented:
See https://github.com/prometheus/prometheus/pull/16773 .

Assuming that PR is now really what we want, we have the blueprint for how to do Kahan summation in native histograms. (Sadly, with the adaptive approach still being better than always just doing the incremental mean calculation, things are not as simple as I was hoping for.)

beorn7 commented:
Sounds great. I'll have a look ASAP. (My review queue is quite manageable these days, but I will have limited availability over the next two weeks.)

beorn7 commented on util/kahansum/kahansum.go:
```diff
@@ -0,0 +1,37 @@
+import "math"
+
+// Inc performs addition of two floating-point numbers using the Kahan summation algorithm.
+func Inc(inc, sum, c float64) (newSum, newC float64) {
```
In lieu if a detailed review just a quick but important note: Have a look at https://github.com/prometheus/prometheus/pull/16895 . It seems plausible to me that inlining and subsequent optimization by the compiler could optimize away the whole Kahan trick, so we should do that here, too.

crush-on-anechka replied:
I've seen that PR. Sure, I'll do this

crush-on-anechka commented:
> And now I'm on vacation… sorry for all the delays. I'll be back in action for good starting August 12.

That's totally fine. Do you think it's a good time to rebase onto the latest main locally to resolve conflicts?

beorn7 commented:
Just a note: This is still on my list. I'm overwhelmed with tasks of various priorities. I'll come to this eventually but cannot provide an ETA right now (but of course other reviewers can always take over and beat me to it).

crush-on-anechka commented:
I’ve rebased and made adjustments, including some bug fixes. After the rebase, two new tests in promql (TestEvaluations/testdata/native_histograms.test) are failing:

```
Line 1596: rate(avg(metric)[10m:])
expected warn annotations but none were found

Line 1601: rate(avg_over_time(metric{id="1"}[10m])[5m:])
expected warn annotations but none were found
```

I haven’t yet investigated the root cause — I’ll continue with this and the CI tests in a couple of days.

beorn7 commented:
I'm finally working on the review. Not quite done yet, but in the meantime, I have an idea about the test failure:

We decided recently to always mark a histogram as gauge if it is the result of a subtraction. As we have always used incremental mean calculation, we had always subtraction involved, and therefore the average was always marked as gauge histogram. I guess this is actually not consistent, because logically, the mean calculation is just an addition with a final division, so it should only result in a gauge if any of the histograms that are averaged are already a gauge.

So I believe your PR has "fixed" this aspect, and the tests were "wrong" before.

However, if we want to be consistent, we have to pay attention that we don't start to mark the outcome as gauge just because we triggered the switch to incremental mean calculation. (I guess in that case we should undo the automatic marking as gauge, if none of the histograms involved are gauges. Maybe there is a pretty solution for that or we just add a final check.)

So far for this breaking news. I'm back to reviewing and hope I'll finish during the next days.

crush-on-anechka commented:
Should we base whether a histogram is marked as a gauge on the operation applied? From what I understand, aside from sum, most aggregations (avg, min, max, quantile, etc.) might be treated as gauge histograms.

beorn7 commented:
> Should we base whether a histogram is marked as a gauge on the operation applied? From what I understand, aside from sum, most aggregations (avg, min, max, quantile, etc.) might be treated as gauge histograms.

Most of the aggregations only make sense on gauges as input, at least in general. (Prometheus proverb: ["First the rate, then aggregate."](https://the-zen-of-prometheus.netlify.app/)). However, there are special cases, where users (hopefully) know what they are doing, and then we don't want to interfere. That's where the sophisticated counter-reset-hint determination for addition is coming from. When it comes to aggregations, there are actually only two aggregations that work with native histograms (NH) as values, `avg` and `sum`. All other aggregations either don't work with NH at all, or they don't modify the values.

`sum` should keep the semantics of the addition. And `avg` should behave as it did what it says on the tin: sum everything up and then divide by the number of samples. However, incremental mean calculation is doing something else behind the scenes, and that creates the problem, because the subtraction involved always sets gauge flavor. There is also another problem, even with direct mean calculation, that the order of the summation will matter for the outcome, given some really pathological combination of counter-reset-hints. So to deal with it properly, we have to determine the counter vs. gauge flavor separately.

Effectively:
- In most cases, all histograms aggregated should be gauges, outcome gauge, all good.
- If all counter reset hints are the same, use that for the outcome (this includes the "all gauge" cause above).
- If there is at least one gauge in the mix, outcome is gauge.
- Otherwise, if there is at least one "unknown" in the mix, outcome is "unknown".
- In any case, if there is at least one "yes, for sure counter reset" and one "no, for sure not a counter reset" in the mix, we annotate with a warning.

The last point is not reliably achieved even with direct mean calculation (because the order of summation matters if we have a suitable mix of counter reset hints), but that might not be that bad, as it is only about the annotation, not about the actual result. But as noticed now, it's annoying that the incremental mean calculation will always create a gauge. We will use incremental mean calculation only rarely with this PR, but in the rare cases that we do, the outcome will be even more surprising. Maybe, to solve things properly and consistently, we should track the counter-reset-hint for the outcome of the aggregation separately in all cases, even for `sum`.

beorn7 reviewed (changes requested):
As far as I can see, this looks very close to the complete solution. Many of my comments are just naming nits and such. A few are more substantial:
- We should have tests that actually fail without the Kahan summation (ideally for `+`, `-`, `avg`/`sum` and `avg_over_time`/`sum_over_time`, to exercise all code paths).
- Reducing the resolution in the "other" histogram does not yet use Kahan summation for merging the buckets. (Bonus points for designing a test to demonstrate that.)
- The case where the base term is zero, but the compensation term is not (which I think is theoretically and maybe even practically possible) is not treated correctly in `compactBuckets`.

And of course, there is the aspect of the correct counter reset hint in aggregations, which is currently wrong or inconsistent upstream. I guess that is something we can solve separately, if you don't like to load it into this already epic PR. (In that case, just modify the currently failing tests with an explanation about what we still have to do.)

Thank you very much for all your efforts with this hairy problem.

beorn7 commented:
FYI: I have filed https://github.com/prometheus/prometheus/issues/17308 for the inconsistency in counter reset handling during aggregations. I will also update the spec to be more clear about it. And I plan to fix this separately before PromCon, so please don't treat it yet in this PR.

crush-on-anechka commented:
I’ve rebased on top of recent `main` with the updates from #17278. I also tried to keep the commit history clean, so it should be easy to go through the new 6 commits individually (starting from defc7f5).

### Merging buckets:

I revisited the logic for merging buckets in the functions `addCustomBucketsWithMismatches`, `kahanReduceResolution`, `zeroCountForLargerThreshold` and `kahanAddBuckets` when both histograms have their own associated compensation terms. Previously, the compensation histograms were merged using plain addition; now, Kahan summation is applied. The new merge flow proceeds in two steps:
```
A, compensationA = kahansum.Inc(B, A, compensationA)
A, compensationA = kahansum.Inc(compensationB, A, compensationA)
```
Thus in the second step compensationB is being added to A using compensationA as the compensation term.

Because of this change, I updated `TestKahanAddWithCompHistogram` in terms of expected compensation histogram. Previously, the test expected the compensation buckets to be merged straightforwardly, matching the new schema and producing predictable compensation bucket values. Now, the test only verifies that the length of the resulting compensation buckets matches the layout of the histogram it’s associated with.

It was a bit tricky in `addCustomBucketsWithMismatches`, specifically in the two `mapBuckets` calls where the origin buckets are merged into `targetBuckets`.
Since histogram `A` carries an associated compensation, I had to modify `mapBuckets` to process `A`'s compensation buckets in parallel during the first call only. So I added a new boolean parameter, `withCompensation`, which is set to `true` when processing `A` (which carries a compensation) and `false` for `B` (which doesn't yet). I tried to avoid making too much noise in this function.

### Go tests:

The test cases `reduce resolution` and `reduce resolution of ‘other’ histogram` in `TestKahanAddWithCompHistogram` check that reducing resolution in either the "receiving" histogram or the "other" one uses Kahan summation and actually fail without it on values `[1e100, -1e100]`.

### Promql tests:

I used values `[1e100, -1e100]` in new and existing tests, where the results should ideally cancel out. These tests fail without Kahan summation, which justifies our efforts, that's the good news.
The bad news is that avg calculation still fail in certain cases.
- `histogram_avg_incremental_2` – With the same input the test fails about 50% of the time for some reason (I used `[1e100, -1e100]` as `h.Sum` values).
- `sum_over_time / avg_over_time` – Fail when values of vastly different magnitudes are mixed, e.g.
`[1, 1e100, 1, 1e50, 2, -1e100, -1e50]`.  
However, if I tweak the order to  
`[1, 1e100, 1, -1e100, 2, 1e50, -1e50]`, the calculation works correctly.

I also noticed that the following query uses plain addition, so it doesn’t handle values like `[1e100, -1e100]` correctly:
`eval instant at 10m sum(histogram_sum{idx="0"} + ignoring(idx) histogram_sum{idx="1"} + ignoring(idx) histogram_sum{idx="2"} + ignoring(idx) histogram_sum{idx="3"})`
I added an analogous test for floats for consistency:
`eval instant at 1m sum(data{test="ten",point="a"}) + sum(data{test="ten",point="b"}) + sum(data{test="ten",point="c"}) + sum(data{test="ten",point="d"})`
After exploring the call stack for these queries, I couldn’t find an obvious place to integrate the Kahan algorithm or maintain a compensation term.

**I've commented out these failing tests for now**.

From what I understand, Kahan summation mainly helps with accuracy when adding lots of values — it prevents small rounding errors from building up into a significant total error. Handling numbers of vastly different magnitudes seems more like a side effect.
Do you think it makes sense to dig into this further by adding more tests with high-magnitude values and trying to refine the algorithm?

### Using KahanAdd:

It's worth noting that `KahanAdd` modifies the compensation histogram `c` in-place when it's non-nil, but if `c` is nil, a new compensation histogram is created inside `KahanAdd` so the original `c` variable remains unchanged and the caller must use the returned value instead. I mentioned this in `KahanAdd` description, but not sure if that's the desired behavior. Previously I ignored the returned value in certain cases and that caused issues. Now I consistently use the returned value from `KahanAdd` in all calls.

### KahanSub:

Is there any use for it? Since we don't use subtraction in average calculation anymore, `KahanSub` is not used anywhere for now.

crush-on-anechka commented on model/histogram/generic.go:
```diff
@@ -205,17 +205,22 @@ func (b *baseBucketIterator[BC, IBC]) strippedAt() strippedBucket[BC] {
+func compactBuckets[IBC InternalBucketCount](
+	primaryBuckets []IBC, compBuckets []float64,
+	spans []Span, maxEmptyBuckets int, isDeltaBuckets bool,
+) (updatedPrimaryBuckets []IBC, updatedCompBuckets []float64, updatedSpans []Span) {
```
Done. I reverted it back to `deltaBuckets` since that name is probably more familiar to everyone.

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -479,6 +549,103 @@ func hasCounterResetCollision(a, b *FloatHistogram) bool {
+			otherNegativeSpans, otherNegativeBuckets,
+			other.Schema, h.Schema,
+			false, false,
+		)
```
Done.

crush-on-anechka commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1406,16 +1406,46 @@ eval instant at 10m avg(histogram_sum)
+    {} {{schema:0 count:Inf sum:10.03120467492675 z_bucket:3.56528e+50 z_bucket_w:0.001 buckets:[2.258e+220 2.2580178264e+220 2.6169037689e+220 261827.54331269444] n_buckets:[4004.342521030831 6080.675675179582 451745.57986202446 2035.3483135107433 1444.171911278132]}}
+
+eval instant at 7m avg_over_time(histogram_sum_over_time_incremental[9m:1m])
+    {} {{schema:0 count:3.497116418577895e+307 sum:1.2539005843658437 z_bucket:4.4566e49 z_bucket_w:0.001 buckets:[2.8225e+219 2.822522283e+219 3.271129711125e+219 32728.442914086805] n_buckets:[500.5428151288539 760.0844593974477 56468.19748275306 254.4185391888429 180.5214889097665]}}
```
I modified these tests and added a few more, they now fail without Kahan summation. The bad news is that a few of them still fail even with Kahan summation. I detailed this in my comment in the conversation section.

crush-on-anechka commented on promql/engine.go:
```diff
@@ -3144,20 +3145,23 @@ func vectorElemBinop(op parser.ItemType, lhs, rhs float64, hlhs, hrhs *histogram
 	groupAggrComplete      bool // Used by LIMITK to short-cut series loop when we've reached K elem on every group.
-	incrementalMean        bool // True after reverting to incremental calculation of the mean value.
+	floatIncrementalMean   bool // True after reverting to incremental calculation for float-based mean value.
+	histIncrementalMean    bool // True after reverting to incremental calculation for histogram-based mean value.
```
Fixed.

beorn7 commented:
Sorry. "ASAP" has turned out to not be very soon… Holidays happened, and then I had various medical issues.

Your reasoning above looks great. I think the remaining pitfalls are acceptable for now.

My thoughts so far:
- Yes, Kahan summation is not perfect, i.e. it doesn't turn floating point arithmetic into exact math. The remaining problems you have found might very well be explained by that (rather than by an imperfect or buggy implementation).
- Not using Kahan summation for the plain `+` operator is "fine". We haven't done this for floats so far, and it would have the potential of confusion (because the engine would calculate `a+b+c` differently then if you did `a+b` first, put the result somewhere (e.g. store it with a recording lure), and then do `result+c` separately). And as you have found, it would be quite involved to implement this. Users have never asked for Kahan summation for the plain `+` operator, while the Kahan summation for `sum` and `avg` aggregation came from practical problems. I guess that's because the aggregations will often aggregate a large number of values (easily thousands), while a query with `+` is usually limited to a few (like a handful).
- `KahanSub` might very well be unneeded now. (I will pay attention during the detailed review.)
- I don't think we need to add many more test cases to prove the usefulness of the Kahan summation. I see the test cases more as proof that the implementation is correct (rather than useful).

I'll have a detailed look at the code ASAP (which this time hopefully really means "soon").

beorn7 commented on promql/promqltest/testdata/aggregators.test:
```diff
@@ -687,6 +687,9 @@ load 10s
 eval instant at 1m sum(data{test="ten"})
 	{} 10
 
+# eval instant at 1m sum(data{test="ten",point="a"}) + sum(data{test="ten",point="b"}) + sum(data{test="ten",point="c"}) + sum(data{test="ten",point="d"})
```
Let's add a comment here why this is inactive, explaining that the plain `+` doesn't use Kahan summation. (I would actually remove the `sum` aggregators from the expression as they are not actually aggregating anything. Just adding the individual points with `+` carries the message most clearly.)

beorn7 replied:
I take this back. The `sum` is essentially used to get rid of the labels. Maybe not the most idiomatic way, but fine.

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -473,6 +590,122 @@ func (h *FloatHistogram) Sub(other *FloatHistogram) (res *FloatHistogram, counte
+// c is a histogram holding the Kahan compensation term. It is modified in-place if non-nil.
+// If c is nil, a new compensation histogram is created inside the function. In this case,
+// the caller must use the returned updatedC, because the original c variable is not modified.
+func (h *FloatHistogram) KahanSub(other, c *FloatHistogram) (updatedC *FloatHistogram, counterResetCollision, nhcbBoundsReconciled bool, err error) {
```
removed

crush-on-anechka commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1388,43 +1388,140 @@ clear
+eval instant at 10m avg(histogram_avg_incremental)
+    {} {{schema:0 count:3.497116418577895e+307 sum:1.2539005843658437 z_bucket:4.4566e49 z_bucket_w:0.001 buckets:[2.8225e+219 2.822522283e+219 3.271129711125e+219 32728.442914086805] n_buckets:[500.5428151288539 760.0844593974477 56468.19748275306 254.4185391888429 180.5214889097665]}}
+
+# eval instant at 10m avg(histogram_avg_incremental_2)
```
this one appeared to be tricky, I wrote a detailed commentary on this test in the main conversation above

crush-on-anechka commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1388,43 +1388,140 @@ clear
+eval instant at 6m avg_over_time(histogram_sum_over_time_incremental_4[7m:1m])
+    {} {{schema:0 count:3.9967044783747367e+307 sum:0.9}}
+
+# eval instant at 6m sum_over_time(histogram_sum_over_time_incremental_5[7m:1m])
```
done

crush-on-anechka commented on model/histogram/float_histogram.go:
```diff
@@ -1369,6 +1650,146 @@ func addBuckets(
+						iSpan--
+						iInSpan = int32(spansA[iSpan].Length)
+						spansA[iSpan].Length++
+						// spansC[iSpan].Length++
```
removed

beorn7 commented:
My spontaneous thought about the rounding: This might work out in our (artificial) test cases, where we do expect round numbers, but in general, the rounding can only reduce accuracy rather than increasing it.

It's inevitable that the accuracy of the float operations depends in their order, and we have decided so far to not go down the rabbit hole and start to reorder operations for improved accuracy (see comments), so we have to accept that the accuracy is "randomly" depending on whatever order happens to be used.

crush-on-anechka commented:
You’re right, I overlooked that the main goal here is to improve accuracy. Rounding intermediate results only to make an artificial test pass is not an effective approach.
I’ve removed the rounding path, commented out the failing test, and added an explanation describing why it doesn't work. The changes were squashed into the existing commits and force-pushed.

beorn7 commented:
Great. I'll give this a final look tomorrow, and then we are hopefully and finally done.

My apologies that this took so long, but I think we have learned a lot on the journey.

One final question: I appreciate your effort to keep the commits meaningful for reviewing. However, I would not like to merge the now 40 commits into main as-is. My idea is to just squash them all into one big "add Kahan summation for native histograms" commit. For example, we don't need to have the journey of adding a KahanSub and then removing it again in main. However, if you feel strongly about it, let me know and then we can check if there is a meaningful grouping of commits to do.

crush-on-anechka commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1388,43 +1388,158 @@ clear
+eval instant at 6m avg_over_time(histogram_sum_over_time_4[7m:1m])
+    {} {{schema:0 count:4 sum:1.1857142857142857}}
+
+# eval instant at 6m sum_over_time(histogram_sum_over_time_5[7m:1m])
```
added a very brief explanation with a reference to the detailed one in the 'load' section

crush-on-anechka commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1388,43 +1388,158 @@ clear
+eval instant at 6m avg_over_time(histogram_sum_over_time_incremental_4[7m:1m])
+    {} {{schema:0 count:3.9967044783747367e+307 sum:0.9}}
+
+# eval instant at 6m sum_over_time(histogram_sum_over_time_incremental_5[7m:1m])
```
same: added a brief explanation and a reference to the detailed one in the 'load' section

beorn7 reviewed (approved):
🎉
Thanks again.

I'm currently re-running the tests. The failure seemed to have been a flaky test. Will merge on green.

crush-on-anechka commented:
I received an email saying that the CI had failed, but I decided to leave it until the morning. Happy to have it all done, thank you for all your guidance!
